File: scripts/javascript/process_images.js
const fs = require('fs');
const path = require('path');
const axios = require('axios');
const csv = require('csv-parser');
const { parse } = require('json2csv');

const csvFilePath = 'csv-files/recipes.csv';
const imagesDir = 'recipes-images';
const imageColumn = 'image';

async function downloadWithRetry(url, options, retries = 3, delay = 2000) {
  try {
    return await axios.get(url, options);
  } catch (error) {
    if (retries <= 0) {
      console.error(`Failed to download ${url} after multiple attempts.`);
      throw error;
    }
    console.log(`Retrying download for ${url}... (${retries} attempts left)`);
    await new Promise((r) => setTimeout(r, delay));
    return downloadWithRetry(url, options, retries - 1, delay * 2);
  }
}

async function downloadImage(imageUrl, imagesDir, failedUrlsLog = 'failed_downloads.log') {
  try {
    const urlPath = new URL(imageUrl).pathname;
    const filename = path.basename(urlPath);
    const imagePath = path.join(imagesDir, filename);

    const response = await downloadWithRetry(imageUrl, {
      responseType: 'arraybuffer',
      timeout: 60000,
    });

    fs.writeFileSync(imagePath, response.data);
    console.log(`Downloaded: ${imagePath}`);
    return imagePath;
  } catch (error) {
    console.error(`Error downloading ${imageUrl}: ${error.message}`);
    fs.appendFileSync(failedUrlsLog, `${imageUrl}\n`);
    return null;
  }
}

async function processCsv() {
  const rows = [];
  fs.mkdirSync(imagesDir, { recursive: true });

  // read the CSV file as a stream
  fs.createReadStream(csvFilePath)
    .pipe(csv()) // for each row, parse the CSV into each object (key -> header, value -> cell)
    .on('data', (row) => rows.push(row)) // collect each parsed row (the object) into this array
    .on('end', async () => {
      console.log(`Read ${rows.length} rows from ${csvFilePath}`);

      const promises = rows.map(async (row) => {
        const imageUrl = row[imageColumn];
        const imagePath = await downloadImage(imageUrl, imagesDir);
        if (imagePath) {
          row[imageColumn] = imagePath;
        }
        return row;
      });

      // wait for all downloads to complete
      const updatedRows = await Promise.all(promises);

      // write the updated rows to a the same file
      const csvData = parse(updatedRows);
      fs.writeFileSync(csvFilePath, csvData);
      console.log(`CSV processed and saved to ${csvFilePath}`);
    });
}

processCsv();

File: scripts/python/process_images.py
#!/usr/bin/env python3
import pandas as pd
import requests
import os
import argparse
from urllib.parse import urlparse
import time

def process_csv(input_csv, output_csv, images_dir, image_column):
    os.makedirs(images_dir, exist_ok=True)
    df = pd.read_csv(input_csv)

    # iterate each row:
    for index, row in df.iterrows():
        image_url = row[image_column]
        try:
            # download the image
            print(f"Downloading image {index+1}/{len(df)}: {image_url}")
            response = requests.get(image_url, timeout=30)
            response.raise_for_status()
            # extract filename from the original url
            url_path = urlparse(image_url).path # eg: https://i.ibb.co/NrfMY0Y/mini-bite-beef-bao.webp
            filename = os.path.basename(url_path)  # eg: mini-bite-beef-bao.webp
            # form a new relative path
            image_path = os.path.join(images_dir, filename) # recipes-images/mini-bite-beef-bao.webp
            # store the image info in the new relative path location:
            with open(image_path, 'wb') as f:
                # put the downloaded image data into the file
                f.write(response.content)
            # update the CSV with the new path
            df.at[index, image_column] = image_path
            print(f"Processed image: {image_path}")
            # add delay between downloads
            time.sleep(0.5)
        except Exception as e:
            print(f"Error processing image at row {index+1}: {str(e)}")

    # save the CSV
    df.to_csv(output_csv, index=False)
    print(f"CSV processed and saved to {output_csv}")

def main():
    parser = argparse.ArgumentParser(description='Process CSV to download images')
    parser.add_argument('--input-csv', required=True, help='Path to input CSV')
    parser.add_argument('--output-csv', required=True, help='Path to output CSV')
    parser.add_argument('--images-dir', required=True, help='Directory to save images')
    parser.add_argument('--image-column', required=True, help='Column with image URLs')

    args = parser.parse_args()

    process_csv(args.input_csv, args.output_csv, args.images_dir, args.image_column)

if __name__ == "__main__":
    main()
